File: flutter_app/lib/features/financial_records/models/financial_record.dart
enum Direction {
  owedToMe('OWED_TO_ME'),
  iOwe('I_OWE');

  final String value;
  const Direction(this.value);

  factory Direction.fromJson(String json) {
    return values.firstWhere((e) => e.value == json, orElse: () => Direction.owedToMe);
  }
}

enum Status {
  pending('PENDING'),
  paid('PAID'),
  cancelled('CANCELLED'),
  overdue('OVERDUE');

  final String value;
  const Status(this.value);

  factory Status.fromJson(String json) {
    return values.firstWhere((e) => e.value == json, orElse: () => Status.pending);
  }
}

enum Category {
  rent('RENT'),
  schoolFees('SCHOOL_FEES'),
  salary('SALARY'),
  loan('LOAN'),
  maintenance('MAINTENANCE'),
  utilities('UTILITIES'),
  subscription('SUBSCRIPTION'),
  personal('PERSONAL'),
  business('BUSINESS'),
  other('OTHER');

  final String value;
  const Category(this.value);

  factory Category.fromJson(String json) {
    return values.firstWhere((e) => e.value == json, orElse: () => Category.other);
  }
}

enum RecurrenceType {
  none('NONE'),
  daily('DAILY'),
  weekly('WEEKLY'),
  monthly('MONTHLY'),
  yearly('YEARLY'),
  custom('CUSTOM');

  final String value;
  const RecurrenceType(this.value);

  factory RecurrenceType.fromJson(String json) {
    return values.firstWhere((e) => e.value == json, orElse: () => RecurrenceType.none);
  }
}

class FinancialRecord {
  final String id;
  final Direction direction;
  final String title;
  final String? description;
  final String? personOrOrganization;
  final double amount;
  final String currency;
  final Category category;
  final DateTime? dueDate;
  final Status status;
  final RecurrenceType recurrenceType;
  final String? recurrenceInterval;

  FinancialRecord({
    required this.id,
    required this.direction,
    required this.title,
    this.description,
    this.personOrOrganization,
    required this.amount,
    required this.currency,
    required this.category,
    this.dueDate,
    required this.status,
    required this.recurrenceType,
    this.recurrenceInterval,
  });

  factory FinancialRecord.fromJson(Map<String, dynamic> json) {
    return FinancialRecord(
      id: json['id'],
      direction: Direction.fromJson(json['direction']),
      title: json['title'],
      description: json['description'],
      personOrOrganization: json['person_or_organization'],
      amount: double.parse(json['amount'].toString()),
      currency: json['currency'],
      category: Category.fromJson(json['category']),
      dueDate: json['due_date'] != null ? DateTime.parse(json['due_date']) : null,
      status: Status.fromJson(json['status']),
      recurrenceType: RecurrenceType.fromJson(json['recurrence_type']),
      recurrenceInterval: json['recurrence_interval'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'direction': direction.value,
      'title': title,
      'description': description,
      'person_or_organization': personOrOrganization,
      'amount': amount,
      'currency': currency,
      'category': category.value,
      'due_date': dueDate?.toIso8601String().split('T')[0],
      'status': status.value,
      'recurrence_type': recurrenceType.value,
      'recurrence_interval': recurrenceInterval,
    };
  }
}

File: backend/app/db/models/financial_record.py
import uuid
import enum
from datetime import datetime, timezone
from sqlalchemy import Column, String, Numeric, Enum as SQLEnum, DateTime, ForeignKey, Date
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.database import Base

class DirectionEnum(str, enum.Enum):
    OWED_TO_ME = "OWED_TO_ME"
    I_OWE = "I_OWE"

class CategoryEnum(str, enum.Enum):
    RENT = "RENT"
    SCHOOL_FEES = "SCHOOL_FEES"
    SALARY = "SALARY"
    LOAN = "LOAN"
    MAINTENANCE = "MAINTENANCE"
    UTILITIES = "UTILITIES"
    SUBSCRIPTION = "SUBSCRIPTION"
    PERSONAL = "PERSONAL"
    BUSINESS = "BUSINESS"
    OTHER = "OTHER"

class StatusEnum(str, enum.Enum):
    PENDING = "PENDING"
    PAID = "PAID"
    CANCELLED = "CANCELLED"
    OVERDUE = "OVERDUE"

class RecurrenceTypeEnum(str, enum.Enum):
    NONE = "NONE"
    DAILY = "DAILY"
    WEEKLY = "WEEKLY"
    MONTHLY = "MONTHLY"
    YEARLY = "YEARLY"
    CUSTOM = "CUSTOM"

class FinancialRecord(Base):
    __tablename__ = "financial_records"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    
    direction = Column(SQLEnum(DirectionEnum), nullable=False, index=True)
    title = Column(String, nullable=False)
    description = Column(String, nullable=True)
    person_or_organization = Column(String, nullable=True)
    
    amount = Column(Numeric(14, 2), nullable=False)
    currency = Column(String, nullable=False, default="USD")
    
    category = Column(SQLEnum(CategoryEnum), nullable=False, default=CategoryEnum.OTHER)
    
    due_date = Column(Date, nullable=True, index=True)
    status = Column(SQLEnum(StatusEnum), nullable=False, default=StatusEnum.PENDING, index=True)
    
    recurrence_type = Column(SQLEnum(RecurrenceTypeEnum), nullable=False, default=RecurrenceTypeEnum.NONE)
    recurrence_interval = Column(String, nullable=True) # e.g. details for custom recurrence
    
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    
    user = relationship("User")
    reminders = relationship("Reminder", back_populates="financial_record", cascade="all, delete-orphan")
